### PR TITLE
Ispn 7599/jolokia server part 2

### PR DESCRIPTION
**Pull request by slaskawi**
https://issues.jboss.org/browse/ISPN-7599

This is a second approach to this ticket. Previously this has been merged by @vblagoje  (https://github.com/infinispan/infinispan/pull/5267) and then reverted by @tristantarrant.

So at the moment this PR is more about how we want to implement this (see [this email](http://lists.jboss.org/pipermail/infinispan-dev/2017-November/017938.html) thread). Basically we have two approaches:

1. Put all the bits into main server distribution. This effectively means that this PR is valid.
2. Put all the bits into integration layers. This effectively means close this PR.

@tristantarrant @gustavonalle @ryanemerson @vblagoje - What's your call on this?

### CODE REVIEW

**Comment by gustavonalle** [comment]
+1 for adding it to the server so everyone can benefit from it!

Wondering why it was reverted? From my understanding, it was because it was too late for 9.1...

**Comment by slaskawi** [comment]
Yes, I think that was the reason... @tristantarrant Maybe you remember?

**Comment by gustavonalle** [comment]
re-triggered CI

**Comment by gustavonalle** [comment]
Can we have a test for this?

File: documentation/src/main/asciidoc/server_guide/management.adoc [hunk 1]
```diff
@@ -65,3 +65,10 @@ In domain mode the port will be the port of the Remoting connector for the Serve
 </subsystem>
 
 ----
+
+=== Jolokia
+
+You can also expose JMX beans using link:https://jolokia.org/[Jolokia]. In order to do this, just run the server with
+additional parameter `--jmx`, for example: `./standalone.xml -c cloud.xml --jmx`. Jolokia configuration is stored in
+`agent.conf` file. It is possible to override this file by specifying it after `--jmx` parameter. For example:
+`./standalone.sh -c cloud.xml --jmx my-config.properties`.
```
Comment (gustavonalle) [inline documentation/src/main/asciidoc/server_guide/management.adoc]: Please add a quick way to check if jolokia is up and running, via ```curl http://127.0.0.1:8778/jolokia/version```

File: server/integration/feature-pack/src/main/resources/content/bin/agent.conf [hunk 2]
```diff
@@ -0,0 +1 @@
+jolokia=port=8778
```
Comment (gustavonalle) [inline server/integration/feature-pack/src/main/resources/content/bin/agent.conf]: With this conf, jolokia will bind only to ```127.0.0.1``` which is a good default security-wise. Is this enough for Openshift to operate? 

Also, Wildfly has cmd line flag ```-b``` flag to control the binding of application level services, and ```-bmanagement``` to control binding of the admin console. Wondering if it's worth it to provide a switch to control the jolokia binding or piggyback in the ```-bmanagement``` one, as an easier way of getting it running quickly without providing a custom config via --jmx.

**Comment by gustavonalle** [comment]
LGTM, I've added a few comments meant as optional improvements

File: documentation/src/main/asciidoc/server_guide/management.adoc (hunk 1 above)
Comment (slaskawi) [reply documentation/src/main/asciidoc/server_guide/management.adoc]: Jolokia spits a startup message to the logs... Do we need to mention additional check here?

File: server/integration/feature-pack/src/main/resources/content/bin/agent.conf (hunk 2 above)
Comment (slaskawi) [reply server/integration/feature-pack/src/main/resources/content/bin/agent.conf]: This file needs to be overridden for OpenShift :)

OpenShift needs different settings and certificates. Here's an example: https://github.com/infinispan/infinispan-openshift-templates/pull/6/files#diff-edf8fc0cedef8b07d126d61e0799548aR74

**Comment by slaskawi** [comment]
@gustavonalle Unfortunately there is no test for it at the moment. I can add it if you want...

File: documentation/src/main/asciidoc/server_guide/management.adoc (hunk 1 above)
Comment (gustavonalle) [reply documentation/src/main/asciidoc/server_guide/management.adoc]: It won't hurt to add a simple check :)

**Comment by gustavonalle** [comment]
@slaskawi yes, I'd like a test, something simple as checking the output of http://127.0.0.1:8778/jolokia/version is enough IMO.

**Comment by tristantarrant** [comment]
Both standalone.sh and jolokia.sh specify `/bin/sh` but you are using bashisms (functions):

`./bin/standalone.sh --jmx
./bin/standalone.sh: 3: ./bin/jolokia.sh: function: not found
jboss-modules.jar
./bin/standalone.sh: 6: ./bin/jolokia.sh: Syntax error: "}" unexpected`

Also, you should add some information about the --jmx switch in the standalone.sh help

**Comment by gustavonalle** [comment]
@tristantarrant good points, but I didn't cache the errors you pasted when running locally, and I suppose it also works via docker. How did you come accross them?

**Comment by tristantarrant** [comment]
This fails when /bin/sh is a symbolic link to something which is not bash. For example, on Debian/Ubuntu, /bin/sh is provided by dash.
The jolokia.sh script can be implemented without a function anyway.

**Comment by gustavonalle** [comment]
@tristantarrant gotcha! Nice catch

**Comment by tristantarrant** [comment]
@slaskawi forgot about this ?

**Comment by slaskawi** [comment]
@tristantarrant I'm not sure where to put the code to display help when typing `./standalone.sh help`.

However I put some comments into the file.

File: server/integration/feature-pack/src/main/resources/content/bin/standalone.bat [hunk 3]
```diff
@@ -13,6 +13,7 @@ rem         standalone.bat --debug 9797
 rem By default debug mode is disable.
 set DEBUG_MODE=false
 set DEBUG_PORT=8787
+set USE_JMX_COLLECTORS=fals
```
Comment (tristantarrant) [inline server/integration/feature-pack/src/main/resources/content/bin/standalone.bat]: you meant false ?

**Comment by slaskawi** [comment]
@gustavonalle Unfortunately I have bad info RT tests...

Arquillian constructs a classpath with JBoss Modules and fires it up (see [this piece of the code](https://github.com/wildfly/wildfly-arquillian/blob/master/container-managed/src/main/java/org/jboss/as/arquillian/container/managed/ManagedDeployableContainer.java#L74)). The idea behind adding Agent Bond was to wire it up in the scripts and try not to touch the internals. Unfortunately this made it almost impossible to test in `integrationtests` module. 

I think the best approach would be to setup a CI job in templates repo. This sounds like a better idea than writing OS-dependent tests in Infinispan repo....

**Comment by gustavonalle** [comment]
@slaskawi no prob, we can add a test later

**Comment by galderz** [comment]
Anything else to add here?

**Comment by gustavonalle** [comment]
Merged, thanks @slaskawi !

**Pull request merged**